File: rabin-miller.py
# Miller-Rabin Probabilistic Primality Test.
#
# It's a primality test, an algorithm which determines whether a given number is prime.
# 
# Theory
# 
# 1. Fermat's little theorem states that if p is a prime and 1<=a<p then a^p-1 = 1(mod p)
#
# 2. If p is a prime x^2 = 1(mod p) or(x-1)(x+1) = 0 (mod p), then x = 1 (mod p) or x = -1 (mod p)
#
# 3. If n is an add prime then n-1 is an even number and can be written as 2^s*d. By Fermat's Little Theorem 
#    either a^d = 1 (mod n) or a^2^r*d = -1 (mod n) for some 0<=r<=s-1
#
# 4. The Rabin-Miller primality test is base on contrapositive of the above claim. That is, if we can find an
#    a(witness) such that a^d != 1 (mod n) and a^2^r*d != -1 (mod p) for all 0<=r<=s-1 then a is witness of compositeness
#    of n and we can say n is not prime, otherwise n may be prime.  
#
# 5. We test our number P for some numbers random a and either declare that p is definitely a composite or probably 
#    a prime. 
#
#    The probably that a composite number is returned as prime after k iterations is 1/4^k.
#  
# The Running Time: O(k log 3 n)
#

import random

def modexp(x, y, mod):
 
    sol = 1

    i = 0

    while (1<<i) <= y:

           if (1<<i)&y:

               sol = (sol * x) % mod

           x = (x * x) % mod

           i = i + 1    

    return sol
   
#
# @param n, n > 3, an odd integer to be tested for primality
# @param accuracy, a parameter that determines the accuracy of the test
# @return false, if n is composite, otherwise probably prime
#
def isPrime(n, accuracy):

     # If the number is 2 or 3, then return True  
     if n == 2 or n == 3:
        return True

     # if the number is negative or oven then I have to return False
     if n<=1 or n&1 == 0:
        return False 
     
     # next step we write n-1 as 2^s*d
     s = 0
     m = n - 1

     while m&1 == 0:
          s += 1
          m >>= 1 

     # now we have and s and d as well
     d = (n-1) / (1<<s)
       
     for i in range(1, accuracy + 1):     
 
         # next step we pick a random number between 2 and n-2
         # we call a a witness for compositeness of n, or is called 
         # strong liar when is n is probably prime to a base a. 
         witness = random.randint(2,n-2)
     
         q = modexp(witness, d, n)

         if q == 1 or q == n - 1:
            continue

         for i in range(0, s):

             q = modexp(q, 2, n)

             if q == 1:
                return False

             if q == n-1:
                break

         return False

     # return n is probably prime
     return True  


print isPrime(21, 3)

File: rabin-miller.cpp
#include <iostream.h>
#include <conio.h>
#include <stdlib.h>

//
//  Algorithm Rabbin-Miller Probabilistic Primality Test
// 
//  Input: A number N to be tested and a variable that determines the accuracy of the test.
//
//  Output: 0 if N is definitely composite or 1 if N is probably a prime.
//
//  Write N as 2^s*d
//  For each iteration
//         Pick a random witness in [2,N-2]
//         q = witness^d mod N
//         if q == 1 || q == N-1 got to nextIteration
//         for each i=1,s-1
//             q = q^2 mod N
//             if q == 1 return composite
//             if q == N-1 return nextIteration      
//         return composite
//  return probably prime                      

//prototype functions
int pick_random(int a,int b);


//function that determines modular exponentiation
long modexp(long x,int y,int mod) {

     long sol = 1;

     for(int i = 0; (1<<i) <= y; ++i) {

	 if( ((1<<i)&y) > 0 ) {

	     sol = (sol * x) % mod;
	 }

	 x = (x * x ) % mod;
     }

     return sol;
}


int isWitness(int possibleWitness, int exponent, int d, int p) {


    possibleWitness = modexp(possibleWitness, d, p);

    // if the possible witness is congruent with 1 and -1,
    // then go to next iteration and
    // pick up another possible witness for compositeness of n
    if(possibleWitness == 1 || possibleWitness == (p - 1)){

       return 0;
    }

    //possibleWitness = 2^r(0..s)*d
    for(int r = 0; r < exponent; r++) {

	    possibleWitness = modexp(possibleWitness, 2, p);

	    if(possibleWitness == 1) {

	       //return composite
	       return 1;

	    }

	    if(possibleWitness == (p-1)) {

	       //return nextIteration
	       return 0;
	    }

    }


    //return composite
    return 1;
}

//check whether a number > 3 and odd is probably prime or definitely composite.
int isPrime(int n, int accuracy) {

     //if the given number is 2 or 3, then return 1, because is obvious!
     if(n == 2 || n == 3) {

	return 1;
     }

     //if the given number is - or even, then return 0!
     if(n<=1 || !(n&1)) {

	return 0;
     }


     //first steps we must to write n-1 as 2^s*d, so s=? and d=?
     int s = 0,
	 exponent,
	 d;

     int possibleWitness;

     //step 1, compute exponent of two 2^s
     for(int m=n-1;!(m&1);s++,m >>=1);

     exponent = s;

     //step 2, compute d of 2^s*d that have to be odd
     d = (n-1)/(1<<s);


     for(int i = 1; i <= accuracy; i++) {

	 //pick a random number as possible witness for compositeness of n
	 possibleWitness = pick_random(2, n - 2);

	  if(isWitness(possibleWitness, exponent, d, n)) {

	       //return composite
	       return 0;
	  }

     }


     //probably n is prime
     return 1;

}

int pick_random(int a,int b) {

    randomize();

    int c = (rand()%(b-a)) + a;

  return c;
}

void main(){

     //given a number n>3 and odd
     int n;

     clrscr();

     cout<<"\n\n"<<"Miller-Rabin Probabilistic Primality Test"<<"\n\n";

     cout<<"n=";cin>>n;

     if(isPrime(n,3)){

	cout<<"Is, probably, prime!";

     }else{

	cout<<"Is composite, hence not prime!";
     }

     cout<<"\n\n\nPress any key for exit!";
     getch();

}